File: rabbitmq_stream/AMQP/nodejs/receive_stream.js
#!/usr/bin/env node

var amqp = require('amqplib');

amqp.connect('amqp://localhost').then(function (conn) {
  process.once('SIGINT', function () { conn.close(); });
  return conn.createChannel().then(function (ch) {

    var q = 'my_first_stream';
    // Define the queue stream
    // Mandatory: exclusive: false, durable: true  autoDelete: false
    var ok = ch.assertQueue(q, {
      exclusive: false,
      durable: true,
      autoDelete: false,
      arguments: {
        'x-queue-type': 'stream',
        'x-max-length-bytes': 2_000_000_000
      }
    })

    ch.qos(100); // this is mandatory

    ok = ok.then(function (_qok) {
      return ch.consume(q, function (msg) {
        console.log(" [x] Received '%s'", msg.content.toString());
        ch.ack(msg);
      }, {
        noAck: false,
        arguments: {
          'x-stream-offset': 'first' // here you canspecify the offset: first, last, next....
        }
      },
      );
    });

    return ok.then(function (_consumeOk) {
      console.log(' [*] Waiting for messages. To exit press CTRL+C');
    });
  });
}).catch(console.warn);

File: rabbitmq_stream/AMQP/python/stream/receive_stream.py
#!/usr/bin/env python3
import pika

credentials = pika.PlainCredentials('guest', 'guest')
connection = pika.BlockingConnection(pika.ConnectionParameters(host="localhost", port=5672,
                                                               virtual_host="/",
                                                               credentials=credentials))

q_name = "stream_queue"
channel = connection.channel()

# Mandatory: exclusive: false, durable: true  auto_delete: false
channel.queue_declare(queue=q_name, auto_delete=False, exclusive=False, durable=True,
                      arguments={
                          'x-queue-type': 'stream',  # Mandatory to define stream queue
                          'x-max-length-bytes': 2_000_000_000
                          # Set the queue retention to 2GB else the stream doesn't have any limit
                      })


def callback(ch, method, properties, body):
    print(" [x] %s" % body)
    ch.basic_ack(delivery_tag=method.delivery_tag)


channel.basic_qos(prefetch_count=100)  # mandatory
channel.basic_consume(
    queue=q_name,
    on_message_callback=callback,
    arguments={
        'x-stream-offset': 'first'  # here you can specify the offset: : first, last, next, and timestamp
        # with first start consuming always from the beginning
    },
    auto_ack=False)
channel.start_consuming()
